core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkedEncodedInputStream.java

```diff
@@ -0,0 +1,293 @@
+        chunkStream.write(CRLF);
+        byte[] newChunkData = chunkStream.toByteArray();
+
+        return Chunk.create(new ByteArrayInputStream(newChunkData), newChunkData.length);
+    }
+
+    /**
+     * Create the final chunk, which includes the header, the extensions, the chunk (if applicable), and the trailer
+     */
+    private Chunk getFinalChunk() throws IOException {
+        ByteArrayOutputStream chunkStream = new ByteArrayOutputStream();
+        writeChunk(END, chunkStream);
+        writeTrailers(END, chunkStream);
+        chunkStream.write(CRLF);
+        byte[] newChunkData = chunkStream.toByteArray();
+
+        return Chunk.create(new ByteArrayInputStream(newChunkData), newChunkData.length);
+    }
+
+    private void writeChunk(byte[] chunk, ByteArrayOutputStream outputStream) throws IOException {
+        writeHeader(chunk, outputStream);
+        writeExtensions(chunk, outputStream);
+        outputStream.write(CRLF);
+        outputStream.write(chunk);
+    }
+
+    private void writeHeader(byte[] chunk, ByteArrayOutputStream outputStream) throws IOException {
+        byte[] hdr = header.get(chunk);
+        outputStream.write(hdr);
+    }
+
+    private void writeExtensions(byte[] chunk, ByteArrayOutputStream outputStream) throws IOException {
+        for (ChunkExtension chunkExtension : extensions) {
+            Pair<byte[], byte[]> ext = chunkExtension.get(chunk);
+            outputStream.write((byte) ';');
+            outputStream.write(ext.left());
+            outputStream.write((byte) '=');
+            outputStream.write(ext.right());
+        }
+    }
+
+    private void writeTrailers(byte[] chunk, ByteArrayOutputStream outputStream) throws IOException {
+        for (Trailer trailer : trailers) {
+            Pair<byte[], byte[]> tlr = trailer.get(chunk);
+            outputStream.write(tlr.left());
+            outputStream.write((byte) ':');
+            outputStream.write(tlr.right());
+            outputStream.write(CRLF);
+        }
+    }
+
+    @Override
+    public synchronized void mark(int readlimit) {
+        // TODO: Implement this
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public synchronized void reset() {
+        // TODO: Implement this
+        throw new UnsupportedOperationException();
+    }
+
+    public static Builder builder() {
+        return new BuilderImpl();
+    }
+
+    public interface Builder {
+
+        /**
+         * Set the backing input stream.
+         */
+        Builder inputStream(InputStream inputStream);
+
+        /**
+         * Set the size of chunks from input stream.
+         * The actual size (in bytes) of an encoded chunk depends on the configuration.
+         */
+        Builder chunkSize(int chunkSize);
+
+        /**
+         * Set the header to be used when creating an encoded chunk.
+         * This header will be the first part of an encoded chunk.
+         */
+        Builder header(ChunkHeader header);
```

Comment:
should this be called a ChunkHeaderFactory to make it clear it's not a constant value? Same with extensions and trailers?

Reply:
I think "XyzProvider" would be a better name.

Reply:
I'll rename the classes to '*Provider'.